### PR TITLE
Implement device tracking

### PR DESCRIPTION
**Pull request by TobiasFella**
Left TODO:

- [x] Check the signature on the keys before storing them
- [x] Persistent key storage across client restarts
- [x] Update tracked users list when users join encrypted rooms and/or room starts being encrypted

### CODE REVIEW

**Comment by TobiasFella** [comment]
Currently also still crashes the client when loading if it fails to load the devicekeys stuff from syncdata cache. I'll have a look at that tomorrow

File: lib/connection.cpp [hunk 1]
```diff
@@ -467,6 +479,11 @@ void Connection::Private::completeSetup(const QString& mxId)
     emit q->stateChanged();
     emit q->connected();
     q->reloadCapabilities();
+#ifdef Quotient_E2EE_ENABLED
+    connectSingleShot(q, &Connection::syncDone, q, [=](){
```
Comment (ognarb) [inline lib/connection.cpp:483]: Something like `connectSingleShot(q, &Connection::syncDone, q, &Connection::Private::createDevicesList);` should probably work.

**Comment by TobiasFella** [comment]
The last part (about tracking users for newly encrypted rooms) is not ideal: it actually calls the function for all rooms that are encrypted on startup, even when the encryption is not new. In this case, it does a little bit too much work, but since no new users will be added to the set of outdated users, no job to update the keys is actually started...

Is this good enough or should I look into improving it further?

**Comment by ognarb** [comment]
> The last part (about tracking users for newly encrypted rooms) is not ideal: it actually calls the function for all rooms that are encrypted on startup, even when the encryption is not new. In this case, it does a little bit too much work, but since no new users will be added to the set of outdated users, no job to update the keys is actually started...
> 
> Is this good enough or should I look into improving it further?

I think it's fine for now. We can always improve more later.

File: lib/connection.cpp (hunk 1 above)
Comment (TobiasFella) [reply lib/connection.cpp:485]: AFAIU it does not, since it's part of Connection::Private which is not a QObject

**Comment by TobiasFella** [comment]
At this point i think I'm done with the content. I'll need to test it a bit more tomorrow and it would be good if you could check if it handles all the necessary cases (especially the race-condition thing from the warning in the docs)

**Pull request merged**